lightning-invoice/src/payment.rs

```diff
@@ -244,140 +253,165 @@ where
 		if invoice.amount_milli_satoshis().is_some() {
 			Err(PaymentError::Invoice("amount unexpected"))
 		} else {
-			self.pay_invoice_internal(invoice, Some(amount_msats), 0)
+			self.pay_invoice_using_amount(invoice, Some(amount_msats))
 		}
 	}
 
-	fn pay_invoice_internal(
-		&self, invoice: &Invoice, amount_msats: Option<u64>, retry_count: usize
+	fn pay_invoice_using_amount(
+		&self, invoice: &Invoice, amount_msats: Option<u64>
 	) -> Result<PaymentId, PaymentError> {
 		debug_assert!(invoice.amount_milli_satoshis().is_some() ^ amount_msats.is_some());
+
 		let payment_hash = PaymentHash(invoice.payment_hash().clone().into_inner());
-		if invoice.is_expired() {
-			log_trace!(self.logger, "Invoice expired prior to first send for payment {}", log_bytes!(payment_hash.0));
+		match self.payment_cache.lock().unwrap().entry(payment_hash) {
+			hash_map::Entry::Occupied(_) => return Err(PaymentError::Invoice("payment pending")),
+			hash_map::Entry::Vacant(entry) => entry.insert(0),
+		};
+
+		let payment_secret = Some(invoice.payment_secret().clone());
+		let mut payee = Payee::from_node_id(invoice.recover_payee_pub_key())
+			.with_expiry_time(expiry_time_from_unix_epoch(&invoice).as_secs())
+			.with_route_hints(invoice.route_hints());
+		if let Some(features) = invoice.features() {
+			payee = payee.with_features(features.clone());
+		}
+		let params = RouteParameters {
+			payee,
+			final_value_msat: invoice.amount_milli_satoshis().or(amount_msats).unwrap(),
+			final_cltv_expiry_delta: invoice.min_final_cltv_expiry() as u32,
+		};
+
+		let send_payment = |route: &Route| {
+			self.payer.send_payment(route, payment_hash, &payment_secret)
+		};
+		self.pay_internal(&params, payment_hash, send_payment)
+			.map_err(|e| { self.payment_cache.lock().unwrap().remove(&payment_hash); e })
+	}
+
+	/// Pays `pubkey` an amount using the hash of the given preimage, caching it for later use in
+	/// case a retry is needed.
+	///
+	/// You should ensure that `payment_preimage` is unique and that its `payment_hash` has never
+	/// been paid before. Because [`InvoicePayer`] is stateless no effort is made to do so for you.
+	pub fn pay_pubkey(
+		&self, pubkey: PublicKey, payment_preimage: PaymentPreimage, amount_msats: u64
+	) -> Result<PaymentId, PaymentError> {
+		let payment_hash = PaymentHash(Sha256::hash(&payment_preimage.0).into_inner());
+		match self.payment_cache.lock().unwrap().entry(payment_hash) {
+			hash_map::Entry::Occupied(_) => return Err(PaymentError::Invoice("payment pending")),
+			hash_map::Entry::Vacant(entry) => entry.insert(0),
+		};
+
+		let params = RouteParameters {
+			payee: Payee::for_keysend(pubkey),
+			final_value_msat: amount_msats,
+			final_cltv_expiry_delta: DEFAULT_MIN_FINAL_CLTV_EXPIRY as u32,
```

Comment:
Technically, keysend users are supposed to provide this explicitly with the pubkey. Technically, keysend is also garbage. If we chose not to expose it, we should at least have a buffer here, as written I'm not sure if you can keysend to LDK lol.

Reply:
Makes sense. Added `final_cltv_expiry_delta` as another parameter.